### ISSUE TITLE
dry-types causes dry-system to crash when its above in gemfile

### ISSUE THREAD

**Issue by senid231** [dry-rb/dry-system#247]
## Describe the bug

if `dry/types` appeared in Gemfile or required in script before `dry/system/container` - script will crash.

## To Reproduce

```ruby
# frozen_string_literal: true

require 'bundler/inline'
$stdout.sync = true

gemfile(true) do
  gem "dry-types", "~> 1.5"  # COMMENT HERE TO FIX
  gem "dry-system", "~> 0.25.0"
  # gem "dry-types", "~> 1.5" # UNCOMMENT HERE TO FIX
end

require 'dry/types' # COMMENT HERE TO FIX
require 'dry/system/container'
# require 'dry/types' # UNCOMMENT HERE TO FIX
```

## Expected behavior

executed without an error

## Actual behavior

```
$ ruby /home/senid/bug_reports/dry_system_bug_report.rb
Resolving dependencies...
Using bundler 2.3.22
Using dry-inflector 0.3.0
Using concurrent-ruby 1.1.10
Using dry-container 0.10.1
Using dry-core 0.8.1
Using dry-auto_inject 0.9.0
Using dry-configurable 0.15.0
Using dry-logic 1.2.0
Using dry-system 0.25.0
Using dry-types 1.5.1
/home/senid/.rvm/gems/ruby-3.1.2@bug_report/gems/dry-system-0.25.0/lib/dry/system/container.rb:76:in `<class:Container>': undefined method `setting' for Dry::System::Container:Class (NoMethodError)

      setting :name
      ^^^^^^^
Did you mean?  String
        from /home/senid/.rvm/gems/ruby-3.1.2@bug_report/gems/dry-system-0.25.0/lib/dry/system/container.rb:72:in `<module:System>'
        from /home/senid/.rvm/gems/ruby-3.1.2@bug_report/gems/dry-system-0.25.0/lib/dry/system/container.rb:28:in `<module:Dry>'
        from /home/senid/.rvm/gems/ruby-3.1.2@dream_app/gems/dry-system-0.25.0/lib/dry/system/container.rb:27:in `<top (required)>'
        from /home/senid/bug_reports/dry_system_bug_report.rb:13:in `require'
        from /home/senid/bug_reports/dry_system_bug_report.rb:13:in `<main>'
```

## My environment

- Affects my production application: **NO**
- Ruby version: 3.1.2
- OS: Debian GNU/Linux 11 (bullseye)

## Additional note

but if we put `dry-types` after `dry-system` in Gemfile - all works fine

**Comment by flash-gordon** [comment]
It's not related to dry-types, it's a matter of loading dry-container and dry-configurable in the correct order. dry-system expects dry-configurable to be available, as part of dry-container's API. But if dry-container is loaded _before_ dry-configurable (i.e. inside `require 'dry/types'`) it falls back to using a built-in emulation of dry-configurable that doesn't work for dry-system. Then even if dry-configurable is loaded it won't help :( You can fix the app code with
```ruby
require 'dry/configurable'
# now when dry-container is required by dry-types it'll find dry-configurable loaded
require 'dry/types'
require 'dry/system/container'
```

@solnic I think we should change dry-container from `defined? Dry::Configurable` to `require 'dry/configurable' rescue LoadError`. WDYT?

**Comment by timriley** [comment]
> @solnic I think we should change dry-container from `defined? Dry::Configurable` to `require 'dry/configurable' rescue LoadError`.

+1

**Comment by solnic** [comment]
I'm sorry about this. I just fixed it in dry-container 0.11.0 as suggested.

https://github.com/dry-rb/dry-container/releases/tag/v0.11.0
